static/app/components/forms/compositeSelect.tsx

```diff
@@ -0,0 +1,137 @@
+import {useMemo, useState} from 'react';
+
+import CompactSelect from 'sentry/components/forms/compactSelect';
+import {GeneralSelectValue} from 'sentry/components/forms/selectControl';
+import {valueIsEqual} from 'sentry/utils';
+/**
+ * CompositeSelect simulates independent selectors inside the same dropdown
+ * menu. Each selector is called a "section". The selection value of one
+ * section does not affect the value of the others.
+ */
+type Section<OptionType> = {
+  /**
+   * Text label that will be display on top of the section.
+   */
+  label: string;
+  /**
+   * Selectable options inside the section.
+   */
+  options: OptionType[];
+  /**
+   * Must be a unique indentifying key for the section. This value will be
+   * used in the onChange return value. For example, if there are two
+   * sections, "section1" and "section2", then the onChange callback will be
+   * invoked as onChange({section1: [selected option values], section2:
+   * [selected option values]}).
+   */
+  value: string;
+  defaultValue?: any;
+  /**
+   * Whether the section has multiple (versus) single selection.
+   */
+  multiple?: boolean;
+  onChange?: (value: any) => void;
+};
+
+type ExtendedOptionType = GeneralSelectValue & {
+  selectionMode?: 'multiple' | 'single';
+};
+
+type Props<OptionType> = Omit<
+  React.ComponentProps<typeof CompactSelect>,
+  'multiple' | 'defaultValue' | 'onChange'
+> & {
+  sections: Section<OptionType>[];
+};
+
+/**
+ * Special version of CompactSelect that simulates independent selectors (here
+ * implemented as "sections") within the same dropdown menu.
+ */
+function CompositeSelect<OptionType extends GeneralSelectValue = GeneralSelectValue>({
+  sections,
+  ...props
+}: Props<OptionType>) {
+  const [values, setValues] = useState(sections.map(section => section.defaultValue));
```

Comment:
This has the caveat that if the `sections` prop changes, the values do not get updated. This may require an `useEffect` that updates `values` whenever `sections` changes.

Reply:
Hmm one issue with that is when `sections` changes, the `values` array will reset itself and mess with the change detection function down below - https://github.com/getsentry/sentry/blob/vl/composite-select/static/app/components/forms/compositeSelect.tsx#L132-L135

Reply:
Ideally I think we shouldn't allow people to change the sections array at all (i.e. add or remove sections) cause that would really mess with `values` array. (Changing the `options` arrays inside each section should be allowed though).

Reply:
Can we at least document this caveat so if it does happen, it's more obvious why?

Reply:
Sure I'll add a comment on the `sections` prop!